### ISSUE TITLE
[FEATURE] Hit Selecting + W-Tap Hit Selecting (with integration)

### ISSUE THREAD

**Issue by Coccocoahelper** [CCBlueX/LiquidBounce#2978]
### LiquidBounce Branch

Both

### Describe your feature request.

There are multiple phenomenons which make hit selecting exceedingly useful, so I'll start from that:

1: Due to how combat mechanics work, you can get free hits / combos if you hit at a different time than your opponent, given the correct distance.
2: Hit selecting makes you hit the ground earlier — this may lead to you being faster, which, combined with a sprint resetting technique, will make you deal more knockback than your opponent, potentially allowing you to initiate a combo. In many servers, there is also a Spigot fork that has custom knockback, perhaps including a ground and a sprint multiplier — making the advantage even greater.
3: Minecraft's attack range is based on a vertical beam starting from your eyes, and is not entirely consistent (by that, I mean that players lower than others will have a slightly higher range).
4: Players make lots of tiny calculation mistakes, and hit selecting takes advantage of that — since your opponent is not taking knockback right away, it'll briefly confuse that same opponent's aim, and you might be able to get a combo.

These advantages can (and should) be combined with multiple other methods, such as W-tapping, S-tapping (both variants), block-hitting, and strafing.

Now, for the feature requests:

1: A fully capable hit selecting option in KillAura, that predicts the distance between you and the enemy, and decides whether to attack last or not; that is because there *are* advantages in getting the first hit, for example if you distance yourself from the enemy right after (for 1–2 ticks). This should also counter hit selectors, as per the second video in the additional context part.
2: A smart option in W-Tap, since it wouldn't be appropriate to do in KillAura. This functions by predicting the distance between you and the enemy (including the knockback), and picking the ticks beneficial by W-tapping. This could help to keep combos, and maybe create some!
3: Overall integration with other combat-related settings/modules, such as block-hitting.

Finally, these features could potentially be detected in highly advanced anti-cheats. However, this should not be a concern — the majority of the anti-cheats (Grim, Intave, etc.) do not check for such things, maybe except Polar. Furthermore, it is possible to minorly randomize the hit selecting process to avoid detection — this could be implemented in a manner that hardly sacrifices the benefits gained.

P.S. This is a further contextualized version of #1738, please don't close this; it could possibly be a good resource (even for other clients' developers).

### Additional context

Here are some resources that I used to make this feature request:

- Nearly complete 1.8 PvP guide: https://www.youtube.com/watch?v=X0z7U6hRpLo
- Hit Selecting: https://youtube.com/watch?v=ndHSpA-6bxk
- Hit Selecting counter: https://youtube.com/watch?v=BbkOmpl47Mc

**Comment by superblaubeere27** [comment]
I decided that the client should at some time get such a module.

> Last but not least, these feature could potentially be detected in some advanced anti-cheats. However, this should not be a concern, because most anti-cheats (e.g., Grim, Intave, etc) do not have those checks.

What checks do you mean? How do they work?

And can you contact me on Discord?

**Comment by Coccocoahelper** [comment]
> I decided that the client should at some time get such a module.
> 
> > Last but not least, these feature could potentially be detected in some advanced anti-cheats. However, this should not be a concern, because most anti-cheats (e.g., Grim, Intave, etc) do not have those checks.
> 
> What checks do you mean? How do they work?
> 
> And can you contact me on Discord?

I do not know how they work, but they might be a thing (if not now, then certainly in the future).

Regarding the Discord contacting, add me @ `iamthatonecoder_86521`

**Comment by masterBeans990** [comment]
you should implement tickbase or create new normal timerrange module to this ong 🙏🔥

**Comment by Coccocoahelper** [comment]
> you should implement tickbase or create new normal timerrange module to this ong 🙏🔥

There are many things that could be done, in order to beat clients such as Augustus. However, I am not competent enough to do them.

**Comment by masterBeans990** [comment]
i want to combo like augustus.. 👉👈

**Comment by SuicidalEeper** [comment]
One day

**Comment by supastishn** [comment]
i'm pretty sure w tapping is available in super knockback so no need for that.
i want this module tho so i'll try pring this

**Comment by Coccocoahelper** [comment]
> I'm pretty sure W-tapping is available in SuperKnockback, so there's no need for that. However, I do want this module, so I'll try to create this module

W-tapping is in SuperKnockback, but it fails to account for techniques your opponents may perform, the server knockback, other modules that may be on (lag-based modules like BackTrack, FakeLag, and TickBase, KillAura and/or AutoClicker, etc.).

To be honest, LiquidBounce should include an engine that predicts what you and other players might do. This would be rather time-consuming, but would turn out to be an overpowered feature.

**Comment by supastishn** [comment]
I'm working on this right now. https://github.com/supastishn/lb-hitselect is the repo
Not sure when i'll be done as i gotta balance this with other stuff tho

> To be honest, LiquidBounce should include an engine that predicts what you and other players might do. This would be rather time-consuming, but would turn out to be an overpowered feature.

That does sounds pretty good yea, but itd take a long time and I'm neither willing nor good enough at kotlin to do that kek

You could open an issue for it and see

**Comment by Coccocoahelper** [comment]
> I'm working on this right now: https://github.com/supastishn/lb-hitselect I'm not sure when I'll be done as I have to balance this with other stuff, though.
> 
> > To be honest, LiquidBounce should include an engine that predicts what you and other players might do. This would be rather time-consuming, but would turn out to be an overpowered feature.
> 
> That does sound good, but it'd take a long time and I'm neither willing nor good enough at Kotlin to do that, kek.
> 
> I assume you could open an issue for it, and see whether anyone else is up for the challenge.

You can detect other players hit selecting by checking if they're swinging close to the attack range. This definitely has flaws, but it's a start. Additionally, you should check for the hit delay between you and your opponent, to avoid it from becoming a disadvantage.

**Comment by Coccocoahelper** [comment]
Progression has been made, and my client already has it. For anyone willing to port this to LiquidBounce, [here is the code](https://github.com/LibreBounce/LibreBounce/blob/main/src/main/java/net/ccbluex/liquidbounce/features/module/modules/combat/SmartHit.kt).

Additionally, the issues #5999 and #7207 have some suggestions that ought to be implemented in my test.

**Comment by Arsh-24** [comment]
Damn how come this was suggested a year ago and still not added to LB, but thanks for letting me know about this @Coccocoahelper ill use your code to make my own version xD

**Comment by mrnuser** [comment]
Yes we need it. After two years of clearly blatanting on Hypixel i can say that.

**Comment by SuicidalEeper** [comment]
Any day now

**Comment by Coccocoahelper** [comment]
SmartHit in LibreBounce is now mostly complete (and provides real advantages), so there should be no reason not to port it over to nextgen.

**Comment by Coccocoahelper** [comment]
> This issue is stale because it has been open 60 days with no activity. Remove stale label or comment or this will be closed in 365 days.

...

**Comment by SuicidalEeper** [comment]
I beg

**Comment by Coccocoahelper** [comment]
> This issue is stale because it has been open 60 days with no activity. Remove stale label or comment or this will be closed in 365 days.

...

**Comment by SuicidalEeper** [comment]
Depsair
